lib/dust.js

```diff
@@ -112,41 +112,79 @@
+    if(!nameOrTemplate) {
+      return;
+    }
+    if(typeof nameOrTemplate === 'function' && nameOrTemplate.template) {
+      // Sugar away CommonJS module templates
+      return nameOrTemplate.template;
+    }
+    if(dust.isTemplateFn(nameOrTemplate)) {
+      // Template functions passed directly
+      return nameOrTemplate;
+    }
+    if(loadFromCache !== false) {
+      // Try loading a template with this name from cache
+      return dust.cache[nameOrTemplate];
+    }
+  }
+
   function load(nameOrTemplate, chunk, context) {
     if(!nameOrTemplate) {
       return chunk.setError(new Error('No template or template name provided to render'));
     }
 
-    if(!dust.config.cache) {
+    if(dust.config.cache === false) {
       dust.cache = {};
     }
 
-    var tmpl;
-    if(typeof nameOrTemplate === 'function' && nameOrTemplate.template) {
-      // Sugar away CommonJS module templates
-      tmpl = nameOrTemplate.template;
-    } else if(dust.isTemplateFn(nameOrTemplate)) {
-      // Template functions passed directly
-      tmpl = nameOrTemplate;
-    } else {
-      // Load a template with this name from cache
-      tmpl = dust.cache[nameOrTemplate];
-    }
+    var template = getTemplate(nameOrTemplate);
 
-    if (tmpl) {
-      return tmpl(chunk, Context.wrap(context, tmpl.templateName));
+    if (template) {
+      return template(chunk, Context.wrap(context, template.templateName));
     } else {
       if (dust.onLoad) {
         return chunk.map(function(chunk) {
-          dust.onLoad(nameOrTemplate, function(err, src) {
+          // Alias just so it's easier to read that this would always be a name
+          var name = nameOrTemplate;
+          // Three possible scenarios for a successful callback:
+          //   - `require(nameOrTemplate)(dust); cb()`
+          //   - `src = readFile('src.dust'); cb(null, src)`
+          //   - `compiledTemplate = require(nameOrTemplate)(dust); cb(null, compiledTemplate)`
+          function callback(err, srcOrTemplate) {
+            var template;
             if (err) {
               return chunk.setError(err);
             }
-            if (!dust.cache[nameOrTemplate]) {
-              dust.loadSource(dust.compile(src, nameOrTemplate));
+            // Prefer a template that is passed via callback over the cached version.
+            template = getTemplate(srcOrTemplate, false) || getTemplate(name);
+            if (!template) {
+              // It's a template string, compile it and register under `name`
+              if(dust.compile) {
+                dust.loadSource(dust.compile(srcOrTemplate, name));
+                template = dust.cache[name];
+              } else {
+                return chunk.setError(new Error('Dust compiler not available'));
+              }
             }
-            dust.cache[nameOrTemplate](chunk, Context.wrap(context, nameOrTemplate)).end();
-          });
+            template(chunk, Context.wrap(context, template.templateName)).end();
```

Comment:
To clarify with code: all you have to do is

    template.templateName = "whatever I want instead of foo_fr"
    callback(null, template);

Whereas in the reverse case, this would be impossible:

    if(name === 'foo') {
        tmpl = load('bar');
        callback(null, tmpl); // context.templateName is always foo, I have no way to know that bar was loaded instead
    }